File: 2023/day04/yuri.fs
module Day4

let input = System.IO.File.ReadAllLines "input\day04.txt"

type ScratchCard = { CardNum: int; WinningNumbers: int Set; MyNumbers: int Set; }

let parseNumbers (str: string) =
    str.Trim().Replace("  ", " ").Split(' ') |> List.ofArray |> List.map (fun x -> x.Trim() |> int) |> Set.ofList

let parseScratchCard (lineIndex: int, line: string) : ScratchCard =
    let split = line.[line.IndexOf(':') + 1..].Split('|')
    { CardNum = lineIndex + 1; WinningNumbers = split[0] |> parseNumbers; MyNumbers = split[1] |> parseNumbers; }

let score (scratchCard: ScratchCard): int =
    pown 2 ((Set.intersect scratchCard.WinningNumbers scratchCard.MyNumbers |> Set.count) - 1)

let scratchCards = input |> List.ofArray |> List.indexed |> List.map  parseScratchCard
let part1 = scratchCards |> List.sumBy score
printfn "Part 1: %A" part1

let countMatchingNumbers scratchCard = 
    Set.intersect scratchCard.WinningNumbers scratchCard.MyNumbers |> Set.count

let countTotalOfAllCards cards =
    let matchesPerCard = cards |> List.map countMatchingNumbers
    let amountPerCard = List.init (cards |> List.length) (fun _ -> 1)

    let fld (apc: int list) (idx: int, amount: int) : int list =
        apc 
        |> List.indexed
        |> List.map (fun (index, value) ->
            if index > idx && index <= idx + amount // in the range of indices to be adjusted  
            then value + (1 * apc[idx]) // increase by amount of cards in current index
            else value)

    matchesPerCard
    |> List.indexed
    |> List.fold fld amountPerCard
    |> List.sum

let part2 = countTotalOfAllCards scratchCards 
printfn "Part 2: %A" part2

File: 2025/day07/yuri.cs
var beams = new HashSet<int>();
var splitCount = 0;

foreach (var line in File.ReadLines("input-example.txt"))
{
    if (beams.Count == 0)
    {
        beams.Add(line.IndexOf('S'));
        continue;
    }

    var nextBeams = new HashSet<int>();

    foreach (var x in beams)
    {
        if (line[x] == '^')
        {
            splitCount++;

            nextBeams.Add(x + 1);
            nextBeams.Add(x - 1);
        }
        else
        {
            nextBeams.Add(x);
        }
    }

    beams = nextBeams;
}

Console.WriteLine($"Part 1: {splitCount}");

var lines = File.ReadAllLines("input.txt").ToArray();
var timelineBeams = new long[lines[0].Length];
timelineBeams[lines[0].IndexOf('S')] = 1;

foreach (var line in lines.Skip(1))
{
    for (var x = 0; x < timelineBeams.Length; x++)
    {
        if (line[x] == '^' && timelineBeams[x] > 0)
        {
            timelineBeams[x + 1] += timelineBeams[x];
            timelineBeams[x - 1] += timelineBeams[x];
            timelineBeams[x] = 0;
        }
    }
}

Console.WriteLine($"Part 2: {timelineBeams.Sum()}");